### ISSUE TITLE
Ensure xml:id restriction works with `generate`

### ISSUE THREAD

**Issue by oscarlevin** [PreTeXtBook/pretext-cli#579]
In a recent run of my book, this didn't seem to work correctly.

**Comment by StevenClontz** [comment]
Yeah I think we have hit this issue with the TBIL project - the numbering of images without xml:ids is borked. Thanks @siwelwerd for making the connection.

So an author can avoid the issue for now by giving images xml:ids, but we need to figure out where things are breaking down.

**Comment by rbeezer** [comment]
And there is going to be some major changes w.r..t `@xml:id`  and `@label`, especially with regard to `image` and a child `latex-image` (and similar).  Working on it now.  The goal is to keep this from getting out of hand.

The whole subtree drill was meant to appease people who wanted to   author-compile-debug  quickly, so I'm not going to be too upset of it is not producing production-quality results.

**Comment by oscarlevin** [comment]
In the work to revise `@xml:id` and `@label`, it would be nice to consider how those can interact with generating specific assets.  The new and very well received feature of the CLI to automatically generate those assets that have changed uses the subtree options.  It would be a shame if this wasn't preserved or reimplemented somehow.

**Comment by rbeezer** [comment]
>  It would be a shame if this wasn't preserved or reimplemented somehow.

Nobody suggested any abandonment of any feature.

It has always been the case that images with filenames like `image-67.pdf` are an expedient meant to provide rapis startup and prototyping  for new or throwaway projects.  In-progess work on `@label` is meant to **improve** the situation.

But I am not surprised that the numbering could be a bit off.  Dave Rosoff recently hit a problem with the numbers being off by two.  Once he was given a quick (undocumented() fix, he has not been interested in making further contributions to help debugging the root cause.

**Comment by oscarlevin** [comment]
The issue is likely only showing up on projects that have some images with `xml:id` and some without.  Will look at more.

**Comment by oscarlevin** [comment]
I should have this fixed by the end of the day (after almost completely rewriting the generate method, fun fun fun).  Once it is done, I will release 2.1

**Comment by oscarlevin** [comment]
Okay, I have refactored the generate function in #622, and I *think* this solves any issues that were present.  Tested all the cases I could think of (but did not do exhaustive testing with every possible configuration).  When that pr is merged, we will close this.  If anyone notices strangeness, we can revisit.

**Comment by StevenClontz** [comment]
> Okay, I have refactored the generate function in #622, and I _think_ this solves any issues that were present. Tested all the cases I could think of (but did not do exhaustive testing with every possible configuration). When that pr is merged, we will close this. If anyone notices strangeness, we can revisit.

It'd be good to have written tests for this stuff, even if they are slow (we don't have to run them on every commit).

**Comment by oscarlevin** [comment]
Given how core handles/recommends labels vs xml:ids for images, I think the best bet is to only use the xml:id restriction on a (sub)section level.  Or use the much improved build command that smartly doesn't regenerate everything.
